pages/blog/blog-post.js

```diff
@@ -0,0 +1,301 @@
+      />}
+
+      <PageSection>
+        <div className="offset-lg-2 col-lg-8">
+          <DateContainer>
+            <h5>{date}</h5>
+          </DateContainer>
+          <BlogTitle>{title}</BlogTitle>
+          <BlogMarkdown source={content} />
+          <div className="row">
+            <AuthorContainer className="col-6">
+              <h4>{props.authorText}</h4>
+              {author}
+            </AuthorContainer>
+            <ShareContainer className="col-6">
+              <h4>{props.shareText}</h4>
+
+              <ShareButtonRow>
+                <FacebookShareButton
+                  url={pageUrl}
+                  quote={shareMessage}
+                >
+                  <SocialLink>
+                    <FaFacebook size={24} />
+                  </SocialLink>
+                </FacebookShareButton>
+
+                <TwitterShareButton
+                  url={pageUrl}
+                  title={shareMessage}
+                >
+                  <SocialLink>
+                    <FaTwitter size={24} />
+                  </SocialLink>
+                </TwitterShareButton>
+              </ShareButtonRow>
+
+            </ShareContainer>
+          </div>
+
+        </div>
+      </PageSection>
+
+    </div>
+  )
+}
+
+BlogPostContent.propTypes = {
+  title: PropTypes.string,
+  slug: PropTypes.string,
+  metaDescription: PropTypes.string,
+  date: PropTypes.string,
+  heroImage: PropTypes.shape(imageShape),
+  content: PropTypes.string,
+  authorTeamMember: PropTypes.shape(teamMemberShape),
+  authorExternal: PropTypes.string,
+}
+
+BlogPostContent.defaultProps = {
+  title: '',
+  slug: '',
+  metaDescription: '',
+  date: '',
+  heroImage: null,
+  content: '',
+  authorTeamMember: null,
+  authorExternal: '',
+}
+
+const BlogPost = props => (
+  props.error ?
+    BlogPostError(props) :
+    (
+      <div>
+        {BlogPostContent(props)}
+
+        <PageSection>
+          <div className="offset-lg-2 col-lg-8">
+            <HorizontalRule />
+
+            <ButtonContainer>
+              <Link route={props.previousPostRoute} passHref>
+                <Button type="secondary">{props.previousPostText}</Button>
+              </Link>
+              <Link route={routes.Blog} passHref>
+                <Button type="secondary">{props.blogHomeText}</Button>
+              </Link>
+              <Link route={props.nextPostRoute} passHref>
+                <Button type="secondary">{props.nextPostText}</Button>
+              </Link>
+            </ButtonContainer>
+          </div>
+        </PageSection>
+      </div>
+    )
+)
+
+BlogPost.propTypes = {
+  ...BlogPostContent.propTypes,
+  ...BlogPostError.propTypes,
+  previousPostRoute: PropTypes.string,
+  nextPostRoute: PropTypes.string,
+}
+
+BlogPost.defaultProps = {
+  ...BlogPostContent.defaultProps,
+  ...BlogPostError.defaultProps,
+}
+
+BlogPost.getInitialProps = async function initialProps({ query }) {
+  return Promise.all([
```

Comment:
The logic in this getInitialProps can be put into the `api/blog` file. This `blog-post` file is already long enough as it is.

Reply:
Hmm, I actually feel it belongs here. The blog post page itself knows that it requires both the "Page: Blog Post" content and the "Blog Post" content. It's conceivable that a different component somewhere else will want to only fetch a "Blog Post" without the page template. Therefore, I prefer to keep the parallel fetch in here.

Reply:
Yeah, but the component file we're in is just concerned with presenting the view, not with fetching the data for the view. Fetching the data is done in the `api` folder. Seperation of concerns and all that.

Reply:
I'd disagree. The concerns have been separated. The concern of the component is knowing what data it needs. The concern of the api is knowing how to fetch that data.

Reply:
The thing is that the distinction between "blog post" and "blog post page" seems irrelevant to anything but the API layer, so I would say that the `api` module should abstract away the difference altogether and just offer a single fetching method. Which basically means moving this code to the blog-post API module and only exporting that.

I do agree that the component is responsible for mapping the fetched result to the props it expects. However, the other page components are designed to expect the exact same format as the one returned by the fetching method, so there shouldn't be much to do other than calling the fetching method.

Reply:
Why is it irrelevant to anything but the API layer? On the home page, when we add the blog extracts, we will most likely make a fetch just for the BlogPost type.

If it's a consistency thing, I can change it though. Just doesn't feel right. :)

Reply:
If you put this function using `Promise.all` in the api folder you will still have one function for blogpost, one for blogpage, and the `Promise.all` function that takes care of the underlying behavior of the api call. Like a helper function.
The component should just care about that the data is being fetched, it shouldn't care about **how** it's fetched.
You could argue that it's the components responsibility to transform the fetched data into the expected props, but that's a sort of middlewareish logic that should be done elsewhere than the component file.
We don't need to add the entire project in one file just because we **can**.